src/transformers/models/llama/modeling_llama.py

```diff
@@ -490,8 +490,8 @@ def forward(
         value_states = self.v_proj(hidden_states)
 
         # Flash attention requires the input to have the shape
-        # batch_size x seq_length x head_dim x hidden_dim
-        # therefore we just need to keep the original shape
+        # batch_size x seq_length x num_heads x head_dim
+        # but rotary embeddings require batch_size x num_heads x seq_length x head_dim
         query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
```

Comment:
yea the next comment:
>         # TODO: These transpose are quite inefficient but Flash Attention requires the layout [batch_size, sequence_length, num_heads, head_dim]. We would need to refactor the KV cache
>        # to be able to avoid many of these transpose/reshape/view.

is probably wrong, the cache does not require anything now, it's more the rope function. `rotate_half` is at fault here, we can probably update it